### PR TITLE
Show error pages for not existing routes and not existing clusters

### LINKED ISSUE #316: Handle Errors in Router
During navigation, the Vue router triggers requests to fetch required data from the backend. If an error occurs, this is not handled properly, the navigation succeeds and the user lands on a broken page.
It would be better if we redirect to a error page in this case. We should introduce several generic error pages (e.g. 404 not found, 500 server error, 403 forbidden etc.).

### LINKED ISSUE #741: Non-existent cluster is shown as creation pending
**What happened**:

I opened a link to a deleted cluster and saw this here instead of an error message:

Screenshot 2020-07-21 at 15 12 27

**What you expected to happen**:
An error message would be nice. ;-)

**How to reproduce it (as minimally and precisely as possible)**:
Just construct a cluster dashboard link for a cluster that doesn't exist.

**Anything else we need to know**:
Nope.

**Environment**:
`v1.43.1`

### PR DESCRIPTION
**Pull request by holgerkoser**
**What this PR does / why we need it**:
This PR adds fallback routes to the `vue-router` which fixes #316. 

If a project does not exist it shows a project not found page.

It also handles subscription errors for not existing or deleted clusters. In this case a more specific error is shown. This fixes #741.

A path pointing to a namespace e.g `/namespace/garden-core` is redirected to `/namespace/garden-core/shoots` and `/namespaces` is redirected to the shoot list of the default namespace.

Unexpected errors during routing will lead to a redirect to adefault error page.

**Which issue(s) this PR fixes**:
Fixes #741 and fixes #316

**Special notes for your reviewer**:

**Release note**:

```improvement user
Always get an error page in case of not found errors
```

### CODE REVIEW

File: frontend/src/store/index.js [hunk 1]
```diff
@@ -866,6 +866,9 @@ const getters = {
   canGetProjectTerminalShortcuts (state, getters) {
     return getters.canGetSecrets
   },
+  canManageProjectTerminalShortcuts (state, getters) {
```
Comment (holgerkoser) [inline frontend/src/store/index.js:869]:
```suggestion
  canUseProjectTerminalShortcuts (state, getters) {
```

File: frontend/src/views/ShootItemPlaceholder.vue [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  methods: {
+    ...mapActions([
+      'subscribeShoot',
+      'subscribeComments',
+      'unsubscribeComments',
+      'fetchInfrastructureSecrets',
+      'fetchProjectTerminalShortcuts'
+    ]),
+    async load (to, from) {
+      if (!this.loading) {
```
Comment (holgerkoser) [inline frontend/src/views/ShootItemPlaceholder.vue:58]: Remove this check if possible.

File: frontend/src/views/NotFound.vue [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  },
+  computed: {
+    ...mapState([
+      'namespace'
+    ]),
+    ...mapGetters([
+      'namespaces'
+    ]),
+    defaultNamespace () {
```
Comment (petersutter) [inline frontend/src/views/NotFound.vue:52]: use defaultNamespace getter

File: frontend/src/router/routes.js [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+   * @prop {string}  [toRouteName]              - It is possible to set a default child route for a top level item (like the PlaceholderComponent)
+   * @prop {string}  [title]                    - Main menu title
+   * @prop {string}  [icon]                     - Main menu icon
+   * @prop {RouteFn} [breadcrumbText]           - Property or function that returns the breadcrumb title
+   * @prop {Tab[]}   [tabs]                     - Determines the tabs to displayed in the main toolbar extenstion slot
+   */
+
+function loginRoute (context, path) {
+  return {
+    path: '/login',
```
Comment (petersutter) [inline frontend/src/router/routes.js:172]: use path param

File: frontend/src/router/routes.js [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    component: Login,
+    meta: {
+      public: true
+    }
+  }
+}
+
+function errorRoute (context, path) {
+  return {
+    path: '/error',
```
Comment (petersutter) [inline frontend/src/router/routes.js:183]: use path param

File: frontend/src/router/routes.js [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+      namespaced: false,
+      projectScope: false,
+      breadcrumbs: accountBreadcrumbs
+    },
+    beforeEnter (to, from, next) {
+      const namespace = state.namespace || getters.defaultNamespace
+      if (!to.query.namespace && namespace) {
+        return next({
+          name: 'Account',
+          query: { namespace }
```
Comment (petersutter) [inline frontend/src/router/routes.js:246]: merge query

File: frontend/src/views/ShootItem.vue [hunk 7]
```diff
@@ -64,6 +64,7 @@ export default {
     }
   },
   mounted () {
+    console.log('terminalSplitpanes')
```
Comment (petersutter) [inline frontend/src/views/ShootItem.vue:67]: rm

File: frontend/src/store/index.js [hunk 8]
```diff
@@ -968,9 +974,29 @@ const actions = {
   },
   subscribeShoot ({ dispatch, commit }, { name, namespace }) {
     return dispatch('shoots/clearAll')
-      .then(() => EmitterWrapper.shootEmitter.subscribeShoot({ name, namespace }))
-      .catch(err => {
-        dispatch('setError', err)
+      .then(() => {
```
Comment (petersutter) [inline frontend/src/store/index.js:977]: async await

File: frontend/src/store/index.js [hunk 9]
```diff
@@ -968,9 +974,29 @@ const actions = {
...
+        return new Promise((resolve, reject) => {
+          EmitterWrapper.shootEmitter.subscribeShoot({ name, namespace })
+          const key = `${name}_${namespace}`
+          const unsubscribe = store.subscribe(({ type, payload }, state) => {
```
Comment (petersutter) [inline frontend/src/store/index.js:981]: add timeout

File: frontend/src/store/index.js [hunk 10]
```diff
@@ -986,18 +1012,30 @@ const actions = {
       })
   },
   subscribeShoots ({ dispatch, commit, state }) {
-    return EmitterWrapper.shootsEmitter.subscribeShoots({ namespace: state.namespace, filter: getFilterValue(state) })
+    return new Promise((resolve, reject) => {
```
Comment (petersutter) [inline frontend/src/store/index.js:1015]: async/await + ignore error in catch

File: frontend/src/store/index.js [hunk 11]
```diff
@@ -1143,20 +1181,18 @@ const actions = {
...
-      commit('SET_KUBECONFIG_DATA', data)
-    }
-  },
-  async fetchProjectTerminalShortcuts ({ commit }) {
-    if (store.state.projectTerminalShortcuts && store.state.projectTerminalShortcuts.namespace === store.state.namespace) {
-      return
+    const { data } = await getKubeconfigData()
+    commit('SET_KUBECONFIG_DATA', data)
+  },
+  async fetchProjectTerminalShortcuts ({ commit, state }) {
```
Comment (petersutter) [inline frontend/src/store/index.js:1187]: ensureTerminalShortcutsLoaded

File: frontend/src/store/index.js [hunk 12]
```diff
@@ -1200,6 +1236,13 @@ const actions = {
     commit('SET_WEBSOCKETCONNECTIONERROR', null)
     return state.websocketConnectionError
   },
+  setSubscriptionError ({ dispatch }, value) {
+    const { kind, code, message } = value
+    if (kind === 'shoot') {
+      return dispatch('shoots/setSubscriptionError', { code, message })
+    }
+    return dispatch('setError', value)
```
Comment (petersutter) [inline frontend/src/store/index.js:1244]: why kind === 'shoot' behaves different

File: frontend/src/store/index.js (hunk 1 above)
Comment (holgerkoser) [reply frontend/src/store/index.js]: ✔️

File: frontend/src/views/NotFound.vue (hunk 3 above)
Comment (holgerkoser) [reply frontend/src/views/NotFound.vue:52]: ✔️

File: frontend/src/router/routes.js (hunk 5 above)
Comment (holgerkoser) [reply frontend/src/router/routes.js]: ✔️

File: frontend/src/views/ShootItemPlaceholder.vue (hunk 2 above)
Comment (holgerkoser) [reply frontend/src/views/ShootItemPlaceholder.vue:58]: Removed the check. It has never been called twice in all cases.

File: frontend/src/router/routes.js (hunk 6 above)
Comment (holgerkoser) [reply frontend/src/router/routes.js:246]: ✔️

File: frontend/src/views/ShootItem.vue (hunk 7 above)
Comment (holgerkoser) [reply frontend/src/views/ShootItem.vue:67]: ✔️

File: frontend/src/store/index.js (hunk 8 above)
Comment (holgerkoser) [reply frontend/src/store/index.js:977]: ✔️

File: frontend/src/store/index.js (hunk 9 above)
Comment (holgerkoser) [reply frontend/src/store/index.js:981]: added 30 seconds timeout
✔️

File: frontend/src/store/index.js (hunk 10 above)
Comment (holgerkoser) [reply frontend/src/store/index.js:1015]: ✔️

File: frontend/src/store/index.js [hunk 13]
```diff
@@ -972,32 +972,44 @@ const actions = {
...
+      }
+      const unsubscribe = store.subscribe((mutation, state) => {
+        switch (mutation.type) {
+          case 'shoots/SET_SUBSCRIPTION_ERROR':
+            return handleSubscriptionError(mutation, state)
+          case 'shoots/SET_SUBSCRIPTION_DONE':
+            return handleSubscriptionDone(mutation, state)
+        }
       })
+      const timeoutId = setTimeout(handleSubscriptionTimeout, 30 * 1000)
```
Comment (petersutter) [inline frontend/src/store/index.js:1010]: 35 seconds (a bit more than the http timeout)

File: frontend/src/views/ShootItemPlaceholder.vue [hunk 14]
```diff
@@ -54,28 +54,26 @@ export default {
...
-            promises.push(this.fetchProjectTerminalShortcuts())
-          }
-          await Promise.all(promises)
-          this.component = 'router-view'
-        } catch (err) {
-          this.component = 'shoot-item-error'
-        } finally {
-          this.loading = false
+    async load ({ name, params }) {
+      this.loading = true
```
Comment (petersutter) [inline frontend/src/views/ShootItemPlaceholder.vue:58]: rm loading

File: frontend/src/utils/Emitter.js [hunk 15]
```diff
@@ -191,16 +191,20 @@ class ShootSubscription extends AbstractSubscription {
 
     /* currently we only throttle NamespacedEvents (for shoots) as for this kind
     * we expect many events coming in in a short period of time */
-    const throttledNsEventEmitter = new ThrottledNamespacedEventEmitter({ emitter: this, wait: 1000 })
+    // const throttledNsEventEmitter = new ThrottledNamespacedEventEmitter({ emitter: this, wait: 1000 })
```
Comment (petersutter) [inline frontend/src/utils/Emitter.js:194]: rm

File: backend/lib/io.js [hunk 16]
```diff
@@ -229,13 +230,18 @@ function setupShootsNamespace (shootsNsp) {
...
-          message: 'Failed to fetch cluster'
-        })
+      } catch (err) {
+        let { code = 500, message = 'Failed to fetch cluster' } = err
+        if (err instanceof HTTPError) {
+          const response = err.response
+          code = response.statusCode
+          if (response.body) {
+            code = response.body.code
+            message = response.body.message
```
Comment (petersutter) [inline backend/lib/io.js:240]: check if needed

File: frontend/src/views/Error.vue [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+      fromRoute: undefined
+    }
+  },
+  computed: {
+    message () {
+      return get(this.$store, 'alert.message')
+    },
+    type () {
+      return get(this.$store, 'alert.type', 'Error').toUpperCase()
+    }
```
Comment (petersutter) [inline frontend/src/views/Error.vue:52]: rm, currently unused

File: frontend/src/views/Error.vue [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+  name: 'router-error',
+  data () {
+    return {
+      fromRoute: undefined
+    }
+  },
+  computed: {
+    message () {
+      return get(this.$store, 'alert.message')
+    },
```
Comment (petersutter) [inline frontend/src/views/Error.vue:49]: message and code... as property

File: frontend/src/views/Home.vue [hunk 19]
```diff
@@ -76,7 +77,7 @@ export default {
     }
   },
   mounted () {
-    if (this.$route.path === '/namespace/+') {
+    if (get(this.$route.query, 'projectDialog') === 'true') {
```
Comment (petersutter) [inline frontend/src/views/Home.vue:80]: add route with `+` and redirect as it is used by the `vscode-gardener-tools`

File: backend/lib/io.js (hunk 16 above)
Comment (holgerkoser) [reply backend/lib/io.js:240]: ✔️

File: frontend/src/store/index.js (hunk 11 above)
Comment (holgerkoser) [reply frontend/src/store/index.js:1187]: ✔️

File: frontend/src/utils/Emitter.js (hunk 15 above)
Comment (holgerkoser) [reply frontend/src/utils/Emitter.js:194]: ✔️

File: frontend/src/views/Error.vue (hunk 18 above)
Comment (holgerkoser) [reply frontend/src/views/Error.vue:49]: ✔️

File: frontend/src/views/Error.vue (hunk 17 above)
Comment (holgerkoser) [reply frontend/src/views/Error.vue:52]: ✔️

File: frontend/src/views/Home.vue (hunk 19 above)
Comment (holgerkoser) [reply frontend/src/views/Home.vue:80]: ✔️

File: frontend/src/store/index.js (hunk 13 above)
Comment (holgerkoser) [reply frontend/src/store/index.js]: ✔️

File: frontend/src/views/ShootItemPlaceholder.vue (hunk 14 above)
Comment (holgerkoser) [reply frontend/src/views/ShootItemPlaceholder.vue:58]: ✔️

File: frontend/src/store/index.js (hunk 12 above)
Comment (holgerkoser) [reply frontend/src/store/index.js]: ✔️

File: backend/lib/io.js [hunk 20]
```diff
@@ -173,6 +173,31 @@ async function subscribeShootsAdmin ({ socket, user, namespaces, filter }) {
...
 }
 
+async function subscribeShoot (socket, { namespace, name}) {
+  leaveShootsAndShootRoom(socket)
+  
+  const kind = 'shoot'
+  const user = getUserFromSocket(socket)
+  const client = user.client
+  try {
+    const shoot = await client['core.gardener.cloud'].shoots.get(namespace, name)
```
Comment (holgerkoser) [inline backend/lib/io.js:183]: Check if user is in project. Join room before list.

File: backend/lib/io.js [hunk 21]
```diff
@@ -173,6 +173,31 @@ async function subscribeShootsAdmin ({ socket, user, namespaces, filter }) {
...
+    joinRoom(socket, `shoot_${namespace}_${name}`)
```
Comment (holgerkoser) [inline backend/lib/io.js:184]: Leave Room in case of an error

File: frontend/src/views/Error.vue [hunk 22]
```diff
@@ -17,73 +17,32 @@ limitations under the License.
...
-        this.$router.push({
+    async goHome () {
+      try {
+        await this.$router.push({
           name: 'Home'
         })
-      } else {
-        this.$router.back()
-      }
+      } catch (err) { /* ignore error */ }
```
Comment (petersutter) [inline frontend/src/views/Error.vue:44]: add comment why we ignore the error

File: frontend/src/views/ProjectNotFound.vue [hunk 23]
```diff
@@ -14,46 +14,30 @@ See the License for the specific language governing permissions and
...
-        <v-btn dark color="cyan darken-2" @click="goBack">
-          Get me out of here
-        </v-btn>
-      </v-col>
-    </v-row>
-  </v-container>
+  <g-error
+    code="404"
+    text="Project not found"
+    message="The project you are looking for doesn't exist or an other error occured!"
```
Comment (petersutter) [inline frontend/src/views/ProjectNotFound.vue:20]: another or an other?

Comment (holgerkoser) [reply frontend/src/views/ProjectNotFound.vue:20]: It is not another error. It is an error other than not found.

File: frontend/src/views/Error.vue (hunk 22 above)
Comment (holgerkoser) [reply frontend/src/views/Error.vue]: ✔️

**Review by petersutter — APPROVED** [review]
/lgtm

**Review by petersutter — APPROVED** [review]
/lgtm

**Pull request merged**